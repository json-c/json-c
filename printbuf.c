/*
 * $Id: printbuf.c,v 1.5 2006/01/26 02:16:28 mclark Exp $
 *
 * Copyright (c) 2004, 2005 Metaparadigm Pte. Ltd.
 * Michael Clark <michael@metaparadigm.com>
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See COPYING for details.
 *
 *
 * Copyright (c) 2008-2009 Yahoo! Inc.  All rights reserved.
 * The copyrights to the contents of this file are licensed under the MIT License
 * (http://www.opensource.org/licenses/mit-license.php)
 */

#include "config.h"

#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_STDARG_H
#include <stdarg.h>
#else /* !HAVE_STDARG_H */
#error Not enough var arg support!
#endif /* HAVE_STDARG_H */

#include "debug.h"
#include "printbuf.h"
#include "snprintf_compat.h"
#include "vasprintf_compat.h"

static int printbuf_extend(struct printbuf *p, int min_size);

struct printbuf *printbuf_new(void)
{
	struct printbuf *p;

	p = (struct printbuf *)calloc(1, sizeof(struct printbuf));
	if (!p)
		return NULL;
	p->size = 32;
	p->bpos = 0;
	if (!(p->buf = (char *)malloc(p->size)))
	{
		free(p);
		return NULL;
	}
	p->buf[0] = '\0';
	return p;
}

/**
 * Extend the buffer p so it has a size of at least min_size.
 *
 * If the current size is large enough, nothing is changed.
 *
 * Note: this does not check the available space!  The caller
 *  is responsible for performing those calculations.
 */
static int printbuf_extend(struct printbuf *p, int min_size)
{
	char *t;
	int new_size;

	if (p->size >= min_size)
		return 0;
	/* Prevent signed integer overflows with large buffers. */
	if (min_size > INT_MAX - 8)
		return -1;
	if (p->size > INT_MAX / 2)
		new_size = min_size + 8;
	else {
		new_size = p->size * 2;
		if (new_size < min_size + 8)
			new_size = min_size + 8;
	}
#ifdef PRINTBUF_DEBUG
	MC_DEBUG("printbuf_memappend: realloc "
	         "bpos=%d min_size=%d old_size=%d new_size=%d\n",
	         p->bpos, min_size, p->size, new_size);
#endif /* PRINTBUF_DEBUG */
	if (!(t = (char *)realloc(p->buf, new_size)))
		return -1;
	p->size = new_size;
	p->buf = t;
	return 0;
}

int printbuf_memappend(struct printbuf *p, const char *buf, int size)
{
	/* Prevent signed integer overflows with large buffers. */
	if (size > INT_MAX - p->bpos - 1)
		return -1;
	if (p->size <= p->bpos + size + 1)
	{
		if (printbuf_extend(p, p->bpos + size + 1) < 0)
			return -1;
	}
	memcpy(p->buf + p->bpos, buf, size);
	p->bpos += size;
	p->buf[p->bpos] = '\0';
	return size;
}

int printbuf_memset(struct printbuf *pb, int offset, int charvalue, int len)
{
	int size_needed;

	if (offset == -1)
		offset = pb->bpos;
	/* Prevent signed integer overflows with large buffers. */
	if (len > INT_MAX - offset)
		return -1;
	size_needed = offset + len;
	if (pb->size < size_needed)
	{
		if (printbuf_extend(pb, size_needed) < 0)
			return -1;
	}

	memset(pb->buf + offset, charvalue, len);
	if (pb->bpos < size_needed)
		pb->bpos = size_needed;

	return 0;
}

int sprintbuf(struct printbuf *p, const char *msg, ...)
{
	va_list ap;
	char *t;
	int size;
	char buf[128];

	/* user stack buffer first */
	va_start(ap, msg);
	size = vsnprintf(buf, 128, msg, ap);
	va_end(ap);
	/* if string is greater than stack buffer, then use dynamic string
	 * with vasprintf.  Note: some implementation of vsnprintf return -1
	 * if output is truncated whereas some return the number of bytes that
	 * would have been written - this code handles both cases.
	 */
	if (size == -1 || size > 127)
	{
		va_start(ap, msg);
		if ((size = vasprintf(&t, msg, ap)) < 0)
		{
			va_end(ap);
			return -1;
		}
		va_end(ap);
		printbuf_memappend(p, t, size);
		free(t);
		return size;
	}
	else
	{
		printbuf_memappend(p, buf, size);
		return size;
	}
}

void printbuf_reset(struct printbuf *p)
{
	p->buf[0] = '\0';
	p->bpos = 0;
}

void printbuf_free(struct printbuf *p)
{
	if (p)
	{
		free(p->buf);
		free(p);
	}
}
