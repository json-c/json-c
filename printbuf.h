/*
 * $Id: printbuf.h,v 1.4 2006/01/26 02:16:28 mclark Exp $
 *
 * Copyright (c) 2004, 2005 Metaparadigm Pte. Ltd.
 * Michael Clark <michael@metaparadigm.com>
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See COPYING for details.
 *
 *
 * Copyright (c) 2008-2009 Yahoo! Inc.  All rights reserved.
 * The copyrights to the contents of this file are licensed under the MIT License
 * (http://www.opensource.org/licenses/mit-license.php)
 */

#ifndef _printbuf_h_
#define _printbuf_h_

#ifdef __cplusplus
extern "C" {
#endif

struct printbuf {
  char *buf;
  int bpos;
  int size;
};

extern struct printbuf*
printbuf_new(void);

/* As an optimization, printbuf_memappend_fast is defined as a macro
 * that handles copying data if the buffer is large enough; otherwise
 * it invokes printbuf_memappend_real() which performs the heavy
 * lifting of realloc()ing the buffer and copying data.
 * Your code should not use printbuf_memappend directly--use
 * printbuf_memappend_fast instead.
 */
extern int
printbuf_memappend(struct printbuf *p, const char *buf, int size);

#define printbuf_memappend_fast(p, bufptr, bufsize)          \
do {                                                         \
  if ((p->size - p->bpos) > bufsize) {                       \
    memcpy(p->buf + p->bpos, (bufptr), bufsize);             \
    p->bpos += bufsize;                                      \
    p->buf[p->bpos]= '\0';                                   \
  } else {  printbuf_memappend(p, (bufptr), bufsize); }      \
} while (0)

/* The following functions provide a printbuf interface where the
 * string terminator '\0' is not always written. This is faster, but
 * the string cannot be used with standard functions while being
 * constructed. To do so, printbuf_terminate_string() must be
 * called first.
 */
extern void
printbuf_memappend_no_nul(struct printbuf *p, const char *buf, int size);
extern void
printbuf_terminate_string(struct printbuf *const p);

#define printbuf_length(p) ((p)->bpos)

/**
 * Set len bytes of the buffer to charvalue, starting at offset offset.
 * Similar to calling memset(x, charvalue, len);
 *
 * The memory allocated for the buffer is extended as necessary.
 *
 * If offset is -1, this starts at the end of the current data in the buffer.
 */
extern int
printbuf_memset(struct printbuf *pb, int offset, int charvalue, int len);

extern int
sprintbuf(struct printbuf *p, const char *msg, ...);

extern void
printbuf_reset(struct printbuf *p);

extern void
printbuf_free(struct printbuf *p);

/**
 * Set initial size allocation for memory when creating strings,
 * as is done for example in json_object_to_json_string(). The
 * default size is 32, which is very conservative. If an app
 * knows it typically deals with larger strings, performance
 * can be improved by setting the initial size to a different
 * number, e.g. 1k. Note that this also means that memory
 * consumption can increase. How far entriely depens on the
 * application and its use of json-c.
 *
 * Note: each time this function is called, the initial size is
 * changed to the given value. Already existing elements are not
 * affected. This function is usually meant to be called just once
 * at start of an application, but there is no harm calling it more
 * than once. Note that the function is NOT thread-safe and must not
 * be called on different threads concurrently.
 *
 * @param size new initial size for printbuf (formatting buffer)
 */
extern void
json_global_set_printbuf_initial_size(int size);
#ifdef __cplusplus
}
#endif

#endif
