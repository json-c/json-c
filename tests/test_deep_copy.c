#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <assert.h>
#include <errno.h>
#include <time.h>

#include "json.h"
#include "printbuf.h"

static void do_benchmark(json_object *src1);

static const char *json_str1 = 
"{"
"    \"glossary\": {"
"        \"title\": \"example glossary\","
"        \"GlossDiv\": {"
"            \"title\": \"S\","
"            \"null_obj\": null, "
"            \"GlossList\": {"
"                \"GlossEntry\": {"
"                    \"ID\": \"SGML\","
"                    \"SortAs\": \"SGML\","
"                    \"GlossTerm\": \"Standard Generalized Markup Language\","
"                    \"Acronym\": \"SGML\","
"                    \"Abbrev\": \"ISO 8879:1986\","
"                    \"GlossDef\": {"
"                        \"para\": \"A meta-markup language, used to create markup languages such as DocBook.\","
"                        \"GlossSeeAlso\": [\"GML\", \"XML\"]"
"                    },"
"                    \"GlossSee\": \"markup\""
"                }"
"            }"
"        }"
"    }"
"}";

static const char *json_str2 =
"{\"menu\": {"
"    \"header\": \"SVG Viewer\","
"    \"items\": ["
"        {\"id\": \"Open\"},"
"        {\"id\": \"OpenNew\", \"label\": \"Open New\"},"
"        null,"
"        {\"id\": \"ZoomIn\", \"label\": \"Zoom In\"},"
"        {\"id\": \"ZoomOut\", \"label\": \"Zoom Out\"},"
"        {\"id\": \"OriginalView\", \"label\": \"Original View\"},"
"        null,"
"        {\"id\": \"Quality\", \"another_null\": null},"
"        {\"id\": \"Pause\"},"
"        {\"id\": \"Mute\"},"
"        null,"
"        {\"id\": \"Find\", \"label\": \"Find...\"},"
"        {\"id\": \"FindAgain\", \"label\": \"Find Again\"},"
"        {\"id\": \"Copy\"},"
"        {\"id\": \"CopyAgain\", \"label\": \"Copy Again\"},"
"        {\"id\": \"CopySVG\", \"label\": \"Copy SVG\"},"
"        {\"id\": \"ViewSVG\", \"label\": \"View SVG\"},"
"        {\"id\": \"ViewSource\", \"label\": \"View Source\"},"
"        {\"id\": \"SaveAs\", \"label\": \"Save As\"},"
"        null,"
"        {\"id\": \"Help\"},"
"        {\"id\": \"About\", \"label\": \"About Adobe CVG Viewer...\"}"
"    ]"
"}}";

static const char *json_str3 =
"{\"menu\": {"
"  \"id\": \"file\","
"  \"value\": \"File\","
"  \"popup\": {"
"    \"menuitem\": ["
"      {\"value\": \"New\", \"onclick\": \"CreateNewDoc()\"},"
"      {\"value\": \"Open\", \"onclick\": \"OpenDoc()\"},"
"      {\"value\": \"Close\", \"onclick\": \"CloseDoc()\"}"
"    ]"
"  }"
"}}";

json_object_to_json_string_fn my_custom_serializer;
int my_custom_serializer(struct json_object *jso, struct printbuf *pb, int level, int flags)
{
	sprintbuf(pb, "OTHER");
	return 0;
}

json_c_shallow_copy_fn my_shallow_copy;
int my_shallow_copy(json_object *src, json_object *parent, const char *key, size_t index, json_object **dst)
{
	int rc;
	rc = json_c_shallow_copy_default(src, parent, key, index, dst);
	if (rc < 0)
		return rc;
	if (key != NULL && strcmp(key, "with_serializer") == 0)
	{
		printf("CALLED: my_shallow_copy on with_serializer object\n");
		void *userdata = json_object_get_userdata(src);
		json_object_set_serializer(*dst, my_custom_serializer, userdata, NULL);
		return 2;
	}
	return rc;
}


int main(int argc, char **argv)
{
	struct json_object *src1, *src2, *src3;
	struct json_object *dst1 = NULL, *dst2 = NULL, *dst3 = NULL;
	int benchmark = 0;

	if (argc > 1 && strcmp(argv[1], "--benchmark") == 0)
	{
		benchmark = 1;
	}

	src1 = json_tokener_parse(json_str1);
	src2 = json_tokener_parse(json_str2);
	src3 = json_tokener_parse(json_str3);

	assert(src1 != NULL);
	assert(src1 != NULL);
	assert(src3 != NULL);

	printf("PASSED - loaded input data\n");

	/* do this 3 times to make sure overwriting it works */
	assert(0 == json_object_deep_copy(src1, &dst1, NULL));
	assert(0 == json_object_deep_copy(src2, &dst2, NULL));
	assert(0 == json_object_deep_copy(src3, &dst3, NULL));

	printf("PASSED - all json_object_deep_copy() returned succesful\n");

	assert(-1 == json_object_deep_copy(src1, &dst1, NULL));
	assert(errno == EINVAL);
	assert(-1 == json_object_deep_copy(src2, &dst2, NULL));
	assert(errno == EINVAL);
	assert(-1 == json_object_deep_copy(src3, &dst3, NULL));
	assert(errno == EINVAL);

	printf("PASSED - all json_object_deep_copy() returned EINVAL for non-null pointer\n");

	assert(1 == json_object_equal(src1, dst1));
	assert(1 == json_object_equal(src2, dst2));
	assert(1 == json_object_equal(src3, dst3));

	printf("PASSED - all json_object_equal() tests returned succesful\n");

	assert(0 == strcmp(json_object_to_json_string_ext(src1, JSON_C_TO_STRING_PRETTY),
	                   json_object_to_json_string_ext(dst1, JSON_C_TO_STRING_PRETTY)));
	assert(0 == strcmp(json_object_to_json_string_ext(src2, JSON_C_TO_STRING_PRETTY),
	                   json_object_to_json_string_ext(dst2, JSON_C_TO_STRING_PRETTY)));
	assert(0 == strcmp(json_object_to_json_string_ext(src3, JSON_C_TO_STRING_PRETTY),
	                   json_object_to_json_string_ext(dst3, JSON_C_TO_STRING_PRETTY)));

	printf("PASSED - comparison of string output\n");

	json_object_get(dst1);
	assert(-1 == json_object_deep_copy(src1, &dst1, NULL));
	assert(errno == EINVAL);
	json_object_put(dst1);

	printf("PASSED - trying to overrwrite an object that has refcount > 1");

	printf("\nPrinting JSON objects for visual inspection\n");
	printf("------------------------------------------------\n");
	printf(" JSON1\n");
	printf("%s\n", json_object_to_json_string_ext(dst1, JSON_C_TO_STRING_PRETTY));
	printf("------------------------------------------------\n");

	printf("------------------------------------------------\n");
	printf(" JSON2\n");
	printf("%s\n", json_object_to_json_string_ext(dst2, JSON_C_TO_STRING_PRETTY));
	printf("------------------------------------------------\n");

	printf("------------------------------------------------\n");
	printf(" JSON3\n");
	printf("------------------------------------------------\n");
	printf("%s\n", json_object_to_json_string_ext(dst3, JSON_C_TO_STRING_PRETTY));
	printf("------------------------------------------------\n");

	json_object_put(dst1);
	json_object_put(dst2);
	json_object_put(dst3);

	printf("\nTesting deep_copy with a custom serializer set\n");
	json_object *with_serializer = json_object_new_string("notemitted");

	char udata[] = "dummy userdata";
	json_object_set_serializer(with_serializer, my_custom_serializer, udata, NULL);
	json_object_object_add(src1, "with_serializer", with_serializer);
	dst1 = NULL;
	/* With a custom serializer in use, a custom shallow_copy function must also be used */
	assert(-1 == json_object_deep_copy(src1, &dst1, NULL));
	assert(0 == json_object_deep_copy(src1, &dst1, my_shallow_copy));

	json_object *dest_with_serializer = json_object_object_get(dst1, "with_serializer");
	assert(dest_with_serializer != NULL);
	char *dst_userdata = json_object_get_userdata(dest_with_serializer);
	assert(strcmp(dst_userdata, "dummy userdata") == 0);

	const char *special_output = json_object_to_json_string(dest_with_serializer);
	assert(strcmp(special_output, "OTHER") == 0);
	printf("\ndeep_copy with custom serializer worked OK.\n");
	json_object_put(dst1);

	if (benchmark)
	{
		do_benchmark(src2);
	}

	json_object_put(src1);
	json_object_put(src2);
	json_object_put(src3);

	return 0;
}

static void do_benchmark(json_object *src2)
{
	json_object *dst2 = NULL;

	int ii;
	/**
	 * The numbers that I got are:
	 * BENCHMARK - 1000000 iterations of 'dst2 = json_tokener_parse(json_object_get_string(src2))' took 71 seconds
	 * BENCHMARK - 1000000 iterations of 'json_object_deep_copy(src2, &dst2, NULL)' took 29 seconds
	 */

	int iterations = 1000000;
	time_t start = time(NULL);

	start = time(NULL);
	for (ii = 0; ii < iterations; ii++) {
		dst2 = json_tokener_parse(json_object_get_string(src2));
		json_object_put(dst2);
	}
	printf("BENCHMARK - %d iterations of 'dst2 = json_tokener_parse(json_object_get_string(src2))' took %d seconds\n", iterations, (int)(time(NULL) - start));

	start = time(NULL);
	dst2 = NULL;
	for (ii = 0; ii < iterations; ii++) {
		json_object_deep_copy(src2, &dst2, NULL);
		json_object_put(dst2);
		dst2 = NULL;
	}
	printf("BENCHMARK - %d iterations of 'json_object_deep_copy(src2, &dst2, NULL)' took %d seconds\n", iterations, (int)(time(NULL) - start));
}

