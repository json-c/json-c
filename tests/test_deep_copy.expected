PASSED - loaded input data
PASSED - all json_object_deep_copy() returned succesful
PASSED - all json_object_deep_copy() returned EINVAL for non-null pointer
PASSED - all json_object_equal() tests returned succesful
PASSED - comparison of string output
PASSED - trying to overrwrite an object that has refcount > 1
Printing JSON objects for visual inspection
------------------------------------------------
 JSON1
{
  "glossary":{
    "title":"example glossary",
    "GlossDiv":{
      "number":16446744073709551615,
      "title":"S",
      "null_obj":null,
      "exixt":false,
      "quantity":20,
      "univalent":19.8,
      "GlossList":{
        "GlossEntry":{
          "ID":"SGML",
          "SortAs":"SGML",
          "GlossTerm":"Standard Generalized Markup Language",
          "Acronym":"SGML",
          "Abbrev":"ISO 8879:1986",
          "GlossDef":{
            "para":"A meta-markup language, used to create markup languages such as DocBook.",
            "GlossSeeAlso":[
              "GML",
              "XML"
            ]
          },
          "GlossSee":"markup"
        }
      }
    }
  }
}
------------------------------------------------
------------------------------------------------
 JSON2
{
  "menu":{
    "header":"SVG Viewer",
    "items":[
      {
        "id":"Open"
      },
      {
        "id":"OpenNew",
        "label":"Open New"
      },
      null,
      {
        "id":"ZoomIn",
        "label":"Zoom In"
      },
      {
        "id":"ZoomOut",
        "label":"Zoom Out"
      },
      {
        "id":"OriginalView",
        "label":"Original View"
      },
      null,
      {
        "id":"Quality",
        "another_null":null
      },
      {
        "id":"Pause"
      },
      {
        "id":"Mute"
      },
      null,
      {
        "id":"Find",
        "label":"Find..."
      },
      {
        "id":"FindAgain",
        "label":"Find Again"
      },
      {
        "id":"Copy"
      },
      {
        "id":"CopyAgain",
        "label":"Copy Again"
      },
      {
        "id":"CopySVG",
        "label":"Copy SVG"
      },
      {
        "id":"ViewSVG",
        "label":"View SVG"
      },
      {
        "id":"ViewSource",
        "label":"View Source"
      },
      {
        "id":"SaveAs",
        "label":"Save As"
      },
      null,
      {
        "id":"Help"
      },
      {
        "id":"About",
        "label":"About Adobe CVG Viewer..."
      }
    ]
  }
}
------------------------------------------------
------------------------------------------------
 JSON3
------------------------------------------------
{
  "menu":{
    "id":"file",
    "value":"File",
    "popup":{
      "menuitem":[
        {
          "value":"New",
          "onclick":"CreateNewDoc()"
        },
        {
          "value":"Open",
          "onclick":"OpenDoc()"
        },
        {
          "value":"Close",
          "onclick":"CloseDoc()"
        }
      ]
    }
  }
}
------------------------------------------------

Testing deep_copy with a custom serializer set
CALLED: my_shallow_copy on with_serializer object

deep_copy with custom serializer worked OK.
